### PR TITLE
Refactor EditablePolygonsLayer into generic EditableGeoJsonLayer

### PR DESCRIPTION
**Pull request by supersonicclay**
`EditableGeoJsonLayer` has the ability to edit any feature in a `FeatureCollection` with any of the following types of geometry:

* Point
* MultiPoint
* LineString
* MultiLineString
* Polygon
* MultiPolygon

_Note: there are a couple known issues with this until we upgrade it to use deck 6.0_

* highlighting of selected feature doesn't always work
* clicking a feature to select it doesn't always work

### CODE REVIEW

File: examples/data/sample-geojson.json [hunk 1]
```diff
@@ -0,0 +1,177 @@
+{
```
Comment (georgios-uber) [inline examples/data/sample-geojson.json]: Can we try to load them with AJAX from https://github.com/uber-common/deck.gl-data/tree/master/examples

File: examples/deck/example.js [hunk 2]
```diff
@@ -1,21 +1,24 @@
...
 
-import testPolygons from '../data/sf-polygons';
+import sampleGeoJson from '../data/sample-geojson.json';
 
 const initialViewport = {
   bearing: 0,
   height: 0,
-  latitude: 37.75,
-  longitude: -122.4,
+  // latitude: 70.13,
```
Comment (georgios-uber) [inline examples/deck/example.js]: Let's try to have some examples in places that are not SF.

File: examples/sf/example.js [hunk 3]
```diff
@@ -281,8 +281,8 @@ export default class Example extends Component<
         }
       ]
     });
-    const nebulaLayers = [segmentsLayer, editableJunctionsLayer, textLayer];
-    const deckLayers = [editablePolygonsLayer];
+    const nebulaLayers = [segmentsLayer, editableJunctionsLayer];
```
Comment (georgios-uber) [inline examples/sf/example.js]: Maybe separate the examples that use `nebulaLayers`.

File: examples/data/sample-geojson.json (hunk 1 above)
Comment (supersonicclay) [reply examples/data/sample-geojson.json]: I could, but then we'd have to deal with async. Is there a benefit to fetching them?

File: examples/deck/example.js (hunk 2 above)
Comment (supersonicclay) [reply examples/deck/example.js]: Agreed, and also more examples that show off more of its capabilities (e.g. editing a single feature as opposed to FeatureCollection). The layer-browser in deck.gl is stellar.

File: examples/sf/example.js (hunk 3 above)
Comment (supersonicclay) [reply examples/sf/example.js]: Agreed. We'll have some cleanup to do with the docs and examples after things solidify.

File: examples/data/sample-geojson.json (hunk 1 above)
Comment (supersonicclay) [reply examples/data/sample-geojson.json]: Good idea. We should add a mixed FeatureCollection there. I pulled the geojson from here: https://github.com/uber/deck.gl/blob/master/examples/layer-browser/data/sample.geo.json

Comment (ibgreen) [reply examples/data/sample-geojson.json]: You are welcome to add more data to the deck.gl-data repository, just check first if the data is under a license. If it is your own hand-crafted GeoJSON then it is of course no problem

File: examples/deck/example.js (hunk 2 above)
Comment (ibgreen) [inline examples/deck/example.js]: Drop comments

File: examples/deck/example.js [hunk 4]
```diff
@@ -87,6 +108,35 @@ export default class Example extends Component<
     this.forceUpdate();
   };
 
+  _renderToolBox() {
+    return (
```
Comment (ibgreen) [inline examples/deck/example.js]: Nit: I think the tool box needs some styling to look its best...

File: examples/deck/example.js [hunk 5]
```diff
@@ -96,15 +146,14 @@ export default class Example extends Component<
       width: window.innerWidth
     };
 
-    const editablePolygonsLayer = new EditablePolygonsLayer({
+    const editableGeoJsonLayer = new EditableGeoJsonLayer({
       data: testFeatures,
       selectedFeatureIndex,
-      pickable: true,
-      isEditing: true,
+      isEditing: this.state.isEditing,
```
Comment (ibgreen) [inline examples/deck/example.js]: What is this flag? Seems too stateful, not declarative... Just call it `editable`?

File: examples/deck/example.js [hunk 6]
```diff
@@ -119,21 +168,23 @@ export default class Example extends Component<
...
+      // Specify the same GeoJsonLayer props
       lineWidthMinPixels: 2,
-      lineWidthMaxPixels: 10,
 
-      // As well as point layer props
-      getPointColor: () => [0x00, 0x20, 0x70, 0xff],
-      pointHighlightColor: [0xff, 0xff, 0xff, 0xff]
+      // Accessors receive an isSelected argument
+      getFillColor: (feature, isSelected) => {
+        return isSelected ? [0x20, 0x40, 0x90, 0xc0] : [0x20, 0x20, 0x20, 0x30];
```
Comment (ibgreen) [inline examples/deck/example.js]: Don't you need some `updateTrigger` to make sure that a change in `isSelected` regenerates the layer's color attributes?

File: src/lib/geojson.js [hunk 7]
```diff
@@ -84,3 +87,154 @@ export function immutablyReplaceCoordinate(
     ...coordinates.slice(indexes[0] + 1)
   ];
 }
+
+export function flattenPositions(geometry) {
+  let positions = [];
+  if (geometry.type === 'Point') {
```
Comment (ibgreen) [inline src/lib/geojson.js]: Nit: I'd personally prefer a `switch (geometry.type)` over a gauntlet of if statements.

File: src/lib/geojson.js [hunk 8]
```diff
@@ -84,3 +87,154 @@ export function immutablyReplaceCoordinate(
...
+      }
+    }
+  }
+
+  return positions;
+}
+
+// TODO: move these to a test file
+
+function assertEquals(expected, actual, message) {
```
Comment (ibgreen) [inline src/lib/geojson.js]: Why are there tests in `src/lib/`? Don't they go into a test dir/separate test file?

File: examples/deck/example.js (hunk 6 above)
Comment (supersonicclay) [reply examples/deck/example.js]: that's in `EditableGeoJsonLayer` itself

Comment (ibgreen) [reply examples/deck/example.js]: Oh I see, isSelected is a parameter. Makes sense then.

**Pull request merged**